hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -640,6 +655,13 @@ private void handleFlushInternal(boolean close)
       }
       LOG.debug("Flush done.");
     }
+
+    if (close) {
+      // When closing, must wait for all flush futures to complete.
+      for (CompletableFuture<Void> flushFuture : allPendingFlushFutures) {
+        flushFuture.get();
+      }
+    }
```

Comment:
Thanks @duongkame for the patch.

> When hsync is called concurrently, lastFlushFuture completion does not guarantee that all the preceeded flushFutures complete.

Does this imply that even for the same `BlockOutputStream` for a key. It is possible that the future can return in a non-sequential order?

Isn't it supposed to be serialized on Raft server on datanode end?

Reply:
> Isn't it supposed to be serialized on Raft server on datanode end?

Yes. Tho the client side closes the connection in another async stage so it will not guarantee the order there.